buku

```diff
@@ -1970,23 +1970,24 @@ class BukuDb:
             Original tag.
         new : list
             Replacement tags.
+        chatty: bool
+            Skip confirmation when set to False.
 
         Returns
         -------
         bool
             True on success, False on failure.
         """
 
-        orig = delim_wrap(orig)
-        newtags = parse_tags(new) if new else DELIM
+        newtags = DELIM.join(sorted(set([s.lower().strip(DELIM) for s in  new])))
 
         if orig == newtags:
             print('Tags are same.')
             return False
 
-        # Remove original tag from DB if new tagset reduces to delimiter
-        if newtags == DELIM:
-            return self.delete_tag_at_index(0, orig)
+        # Remove original tag from DB if new tagset is empty
+        if newtags == '':
```

Comment:
Does Example 31 in README `$ buku --replace 'old tag'` work with this anymore?

Reply:
Also, Example 30.

Reply:
P.S. A pythonic way of checking for the string being empty would be `if not newtags`, BTW

Reply:
So what would be the syntax to replace a tag by 2 tags from the command line?

Reply:
Same as now, I believe? `--replace foo bar,baz`
My suggestions are meant to keep the old CLI behaviour working.

…Alternatively, the CLI syntax can be changed to treat separate parameters as separate tags – in this case, tags containing whitespace would need to be quoted explicitly (i.e. `'foo bar'`  or `"foo bar"`); that's how it's normally done with shell arguments. In this case, `args.replace[1:]` can be passed directly (without preprocessing).

Reply:
Yes, please keep the behaviour same as now.

Reply:
@2ynn the code should be updated to retain existing CLI behaviour (as mentioned [above](https://github.com/jarun/buku/pull/676#discussion_r1130310495)).

To rephrase, there's two options:
1. Do `','.join(new).split(',')` before [processing input](https://github.com/2ynn/buku/blob/put_api_tags/buku#L1982), and do `parse_tags(args.replace[:1])` [in the CLI](https://github.com/2ynn/buku/blob/put_api_tags/buku#L5985)
2. _Replace_ input processing with `parse_tags(','.join(new))`, and do `[' '.join(args.replace[1:])]` in the CLI

The latter should result in simpler code, I believe (though the ‘empty’ value would be `DELIM` again).
…If you pick the former instead, make sure to strip whitespace and filter out empty tag values as well, these.

Reply:
My initial approach was to no alter buku CLI behavior. I personally believe that if `replace_tag()` is intended in that way it should accept a comma-separated string as an argument and not a list. I have modified my code on the API side to parse the input list and create a separated string to be passed to `replace_tags()`. For now this string is wrapped in list as a hack. What is the type of `args.replace[1:]` ?

Reply:
You seem to misunderstand. I didn't mean you should _reverse_ the changes in the function: if it accepts a list, it _should_ treat separate items as such (instead of implicitly joining them with spaces and re-splitting with commas). Alternatively this argument could be made a string, to do away with ambiguity altogether (it'll be split over commas either way).

The problem with your change was that you didn't _update_ CLI code so that it works the same way as it did before. The CLI code accepts a _comma-separated_ tags list which is _split over spaces_ (to avoid needing quotes, I guess?), and passes it into the function _unprocessed_; after changing how the function handles its argument, you need to update the code that _uses_ it.

Like I said before, the simplest way of handling this is to move the `[string]` hack inside of the function – then the CLI-specific space-joining can be done in the CLI code, and any other place where it's called won't have to deal with this CLI-specific stuff (whether the argument is kept as a list of tags, or made a comma-joined string).

Reply:
Ok, thank you for the clarifications. I am going with option 2 but `parse_tags()` is intended to take a list as an argument, passing it a string results in treating the input as a list of characters and adds spaces between each of them. Did you mean `parse_tags([DELIM.join(new)])` ?

Reply:
Yes, basically.

…You don't really need to create a variable just to pass its value into a call immediately, BTW. Especially with that short of an expression.

Reply:
> `[tag.strip(buku.DELIM) for tag in new_tags]`

This doesn't seem necessary, as the commas are handled within the function.

Reply:
from an API perspective, if a user wants to replace an existing tag with tags that happen to have commas in them (e.g. `['tag,2', 'tag,3']) wouldn't the result be unexpected ( `tag`, `1` and `2` in that case) ?

Reply:
* `.strip()` _does not affect the middle of the string_: `',foo,bar,baz,'.strip(',') == 'foo,bar,baz'`
* the result of `parse_tags([','.join([',foo,', ',bar,'])])` is exactly the same as `parse_tags(['foo,bar'])`
* commas are not supported in tags: they're stored as a single string in DB, with `,` as separator
* this is the reason why I suggested making this argument into a string as an alternative solution

If you want to avoid surprises for the user, you can reject requests containing invalid tags. (…Incidentally, have you checked what happens if the original tag argument contains commas? It doesn't seem to be covered by tests.)

> `tags == []`

The pythonic way of checking for emptiness is via `not`, BTW.

Reply:
…A newline is necessary to end a quotation block in Markdown :sweat_smile:

> OK, I confused it with `replace(',', ' ')`

Wouldn't that be even more unexpected?

> I also think returning more detailed error strings than "failure" would be useful.

I suppose you could introduce an optional argument for that. That would require converting values in `response_template` into lambdas.

…Incidentally, `response_ok` and `response_bad` might as well be moved into `response.py` (thus changing their invocations to `response.ok` and `response.bad`).

>  Could this case be covered by a 404 on a non-existing tag or would passing a string containing commas to `search_tag` be problematic?

I just checked and it seems that the current implementation [doesn't account for it at all](https://github.com/jarun/buku/issues/680). I suppose we need to figure out how it _should_ be handled first, and then adjust API behaviour accordingly (…and fix the implementation as well :sweat_smile:).

Reply:
…Speaking of which, the `DELETE` method also should be checked for the commas handling.

Reply:
…Based on the issue discussion, it seems that attempts to replace multiple tags at once should be rejected. I suggest implementing in `replace_tag()` a check for comma in the first argument (handling it the way it does the second argument being identical to the first).

As for the network API, I guess it should produce HTTP 404 for such URIs.